### DISCUSSION TITLE
How to mix the traces from Datadog and OpenTelemetry?

### DISCUSSION THREAD

**Discussion by MacNale** [open-telemetry/opentelemetry-collector-contrib#2724] [Q&A]
Team,

I am having some java services (SERVICE_A) being instrumented with Datadog agents. There are some services (SERVICE_B) which are not being instrumented by Datadog. These are legacy services written in a language which is not supported by any agent. The legacy services produces some traces in proprietary format. I am trying to convert these traces using opentelemetry SDK. This trace has Datadog parent TRACEID from SERVICE_A. 

I am not sure, whether Datadog would be able to take traces from SERVICE_A instrumented using Datadog agent and SERVICE_B traces produced using OpenTelemetry SDK and present as a single transaction.

thanks,
Mac

**Comment by ericmustin** [comment]
👋 Hello , i'm glad you brought this up, I think I can help here. So, for services you're migrating to opentelemetry-<xzy> sdks, but you still want them to connect to datadog-trace-<xyz> sdk instrumented applications, as long as you're adding Propagation to both OpenTelemetry and Datadog using B3 Headers, you should be able to correctly pass trace context between the apps and they'll be connected in the datadog backend

Here's an example of how this would be enabled for opentelemetry-python <=> datadog-trace-rb, for example:

- https://github.com/open-telemetry/opentelemetry-python/blob/0cd037fdd396f68a8080f84b3a71ecea49c1a393/docs/getting-started.rst#configure-your-http-propagator-b3-baggage
- https://docs.datadoghq.com/tracing/setup_overview/custom_instrumentation/ruby/?tab=activespan#b3-headers-extraction-and-injection

Also worth noting that opentelemetry-<xyz> sdk traces would have to go through the opentelemetry-collector+datadogexporter, and datadog-trace-<xyz> sdk traces would have to go through the datadog-agent, but they'd both be connected in your backend/UI.

Does that help?

**Reply by MacNale** [reply]
@ericmustin Thank you for the links.

What happens if the existing application use proprietary context (e.g "Datadog: Datadog"). Will be I able to still fill in the trace from my legacy service and use opentelemetry collector to export to Datadog -

Sudo Java code in the service which builds the traces from legacy service -
```
void buildContext() {
  SpanContext remoteSpanContext = SpanContext.createFromRemoteParent( TRACEID, SPANID,
        TraceFlags.getSampled(), TraceState.getDefault());
  buildTraceAndSendToDataDogThroughCollector(remoteSpanContext);
}

void buildTraceAndSendToDataDogThroughCollector(SpanContext spanContext) {
  OpenTelemetry openTelemetryParent = ExampleConfiguration.initOpenTelemetry();
  Tracer trace = openTelemetryParent.getTracer("LegacyService");
   
   //Read the traces from the file which includes DataDog TraceID.

  Span parentSpan = trace.spanBuilder("Blah")
      .setStartTimestamp(getStartTime(), TimeUnit.MILLISECONDS)
      .setParent(Context.current().with(Span.wrap(spanContext)))
      .startSpan();  

  parentSpan.end(getEndTime(), TimeUnit.MILLISECONDS);
}
```

This is my converter service which uses opentelemetry sdk and emits the traces to OpenTelemetry Collector, so that I can ingest into Datadog. Do you see any problems with this?


Note: I will check with my application team, if they are okay to change the context to "B3" in service_A instrumented by datadog agent.

**Reply by ericmustin** [reply]
@MacNale apologies on the delay i had missed this msg. as far as I know, within an application process, traces generated from dd-trace-java are not interoperable with those from opentelemetry-java. Seperate services can be connected via distributed tracing.

**Reply by MacNale** [reply]
@ericmustin 

These are two independent services.  This is what I want to achieve.

![image](https://user-images.githubusercontent.com/14235758/113615157-1ffdaf00-9619-11eb-8351-271e17c8eea0.png)

**Reply by ericmustin** [reply]
@MacNale Ah, thank you for the clarification. My original answer should be what you're looking for here, apologies as this is not well documented in our docs but is possible, I'll work with the documentation team on updating for this use case.  

A Datadog-instrumented application (service_a), emitting traces to a datadog-agent, can have it's traces connected to an opentelemetry-instrumented application (service_b), emitting traces to an opentelemetry-collector (which would have a datadogexporter configured). It is necessary to enable `B3` trace context propagation (inject) on your datadog application, as well as on your opentelemetry-application. Here is an example for java:

- B3 datadog: https://docs.datadoghq.com/tracing/setup_overview/setup/java?tab=containers#b3-headers-extraction-and-injection
- B3 (multi) opentelemetry: https://github.com/open-telemetry/opentelemetry-java/blob/72db16e58e3e8cd83c213ae1292df698ecd9d553/sdk-extensions/autoconfigure/README.md#propagator

Does that help? If you are having issues with a specific lang i'm happy to try to help troubleshoot further

**Reply by ericmustin** [reply]
Are these services connected via HTTP? my answer here is specific to distributed tracing over HTTP

**Reply by MacNale** [reply]
@ericmustin These are connected with HTTP(SOAP/REST) and IBM MQ.   How do you deal with MQ?  The application is written in COBOL.

**Reply by matthiashaasesourceforge** [reply]
@MacNale Did you solve your problem last days/weeks/years ?
Same problem on my side. But not with COBOL. It's Fortran am mqtt.

**Reply by isaac621** [reply]
@ericmustin Do the datadog-trace- sdk instrumented applications have to enable the otel feature first, as instrumentation here https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/php/otel/#setup? As I see `tracecontext` format ingestion and extaction should be enabled by default in dd-trace library.

**Reply by ericmustin** [reply]
hey @isaac621 . I no longer work at ddog, i left a few years ago. The docs you sent imply you need to set `DD_TRACE_OTEL_ENABLED` to `true`, but again, I would not be the official source here anymore, ddog has done a tremendous amount of work on all things otel since i left (for the better!) so i would check with someone from their team.

@mx-psi , apologies for the ping, would you be able to point these folks in the right direction?

![image](https://github.com/user-attachments/assets/eeb536fa-4a16-46b4-abaf-96d2f6f4951c)

**Reply by mx-psi** [reply]
Thanks Eric for the ping as always! :smile: @isaac621 can you file a support ticket with Datadog?  https://www.datadoghq.com/support/

**Reply by MuhammedEl-Sayed** [reply]
@mx-psi also curious as to the solution here. will open a ticket if necessary, but wondering if it was a simple answer?

**Reply by ericmustin** [reply]
@mx-psi hey do you want me to delete this answer or something? Or is this possible to unselect as the correct answer? This was all in 2021 and Otel's specification, as well as Datadog's support, has changed significantly enough that I'd be shocked if the happy path here was still `B3`? Go away AI here be dragons

**Reply by ericmustin** [reply]
@MuhammedEl-Sayed btw have you reviewed the IBM MQ Docs? And is your question specific to IBM MQ? They have Opentelemetry support now depending on system requirements. https://www.ibm.com/docs/en/instana-observability/1.0.313?topic=mq-tracing-in-premises-environments

**Reply by mx-psi** [reply]
Thanks for the ping! (and hope you are doing well!!) I have unmarked it as the correct answer

**Reply by MuhammedEl-Sayed** [reply]
@ericmustin Not related to IBM! just mixes DD and OTEL traces! Apologies for necroing. I'll take no reply @mx-psi as this not being a straightforward solution.

**Reply by mx-psi** [reply]
> I'll take no reply @mx-psi as this not being a straightforward solution.

I think opening a support case with Datadog is the best next step. You can include a link to this thread to provide context
